WalletWasabi.Fluent/ViewModels/AddWallet/ConnectHardwareWalletViewModel.cs

```diff
@@ -0,0 +1,226 @@
+				.Subscribe(x =>
+				{
+					WalletListVisible = x > 0;
+				});
+		}
+
+		private string WalletName  { get; }
+
+		private WalletManager WalletManager  { get; }
+
+		private HwiClient HwiClient  { get; }
+
+		private Task? DetectionTask  { get; set; }
+
+		private CancellationTokenSource? DetectionCts  { get; set; }
+
+		private CancellationTokenSource? NavigatedToCts { get; set; }
+
+		public HardwareWalletViewModel? SelectedHardwareWallet
+		{
+			get => _selectedHardwareWallet;
+			set => this.RaiseAndSetIfChanged(ref _selectedHardwareWallet, value);
+		}		
+
+		public bool WalletListVisible
+		{
+			get => _walletListVisible;
+			set => this.RaiseAndSetIfChanged(ref _walletListVisible, value);
+		}
+
+		public ObservableCollection<HardwareWalletViewModel> HardwareWallets { get; }
+
+		public ReactiveCommand<string, Unit> OpenBrowserCommand { get; }
+
+		private async void ConnectSelectedHardwareWallet()
+		{
+			if (SelectedHardwareWallet?.HardwareWalletInfo.Fingerprint is null)
+			{
+				return;
+			}
+
+			try
+			{
+				IsBusy = true;
+				await StopDetection();
+
+				var newWallet = await Task.Run(async () =>
+				{
+					var fingerPrint = (HDFingerprint)SelectedHardwareWallet.HardwareWalletInfo.Fingerprint;
+					using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
+					var extPubKey = await HwiClient.GetXpubAsync(SelectedHardwareWallet.HardwareWalletInfo.Model, SelectedHardwareWallet.HardwareWalletInfo.Path, KeyManager.DefaultAccountKeyPath, cts.Token);
+					var path = WalletManager.WalletDirectories.GetWalletFilePaths(WalletName).walletFilePath;
+
+					return KeyManager.CreateNewHardwareWalletWatchOnly(fingerPrint, extPubKey, path);
+				});
+
+				WalletManager.AddWallet(newWallet);
+
+				// Close dialog
+				ClearNavigation();
+			}
+			catch (Exception ex)
+			{
+				// TODO: Notify the user about the error
+				Logger.LogError(ex);
+
+				// Restart detection
+				StartDetection();
+			}
+			finally
+			{
+				IsBusy = false;
+			}
+		}
+
+		private async Task StopDetection()
+		{
+			if (DetectionTask is { } task && DetectionCts is { } cts)
+			{
+				cts.Cancel();
+				await task;
+			}
+		}
+
+		private void StartDetection()
+		{
+			if (NavigatedToCts is { } cts)
+			{
+				DetectionCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token);
+				DetectionTask = HardwareWalletDetectionAsync(DetectionCts);
+			}
+		}
+
+		private async Task HardwareWalletDetectionAsync(CancellationTokenSource detectionCts)
+		{
+			while (!detectionCts.IsCancellationRequested)
+			{
+				var sw = Stopwatch.StartNew();
+
+				try
+				{
+					using var timeoutCts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
+
+					var detectedHardwareWallets = (await HwiClient.EnumerateAsync(timeoutCts.Token)).Select(x => new HardwareWalletViewModel(x)).ToArray();
+					detectionCts.Token.ThrowIfCancellationRequested();
+
+					// The wallets that are already exists in the wallets.
```

Comment:
```suggestion
					// The wallets that already exist in the hardware wallets list.
```

Reply:
This is not correct. That line just gets the wallets (from the detected HW wallets) that are already added to the software to remove later. I would go with something like "The wallets that already exist in the software."

Reply:
Edited.